File: algorithms/cpp/imageSmoother/imageSmoother.cpp
// Source: https://leetcode.com/problems/image-smoother/
// Author: Miao Zhang
// Date:   2021-02-26

class Solution {
public:
    vector<vector<int>> imageSmoother(vector<vector<int>>& M) {
        int m = M.size();
        int n = M[0].size();
        vector<vector<int>> res(m, vector<int>(n));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int cnt = 0;
                for (int dx = -1; dx <= 1; dx++) {
                    for (int dy = -1; dy <= 1; dy++) {
                        int x = i + dx;
                        int y = j + dy;
                        if (x >= 0 && x < m && y >= 0 && y < n) {
                            res[i][j] += M[x][y];
                            cnt++;
                        }
                    }
                }
                res[i][j] /= cnt;
            }
        }
        return res;
    }
};

File: algorithms/python/imageSmoother/imageSmoother.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/image-smoother/
# Author: Miao Zhang
# Date:   2021-02-26

class Solution:
    def imageSmoother(self, M: List[List[int]]) -> List[List[int]]:
        m = len(M)
        n = len(M[0])
        res = [[0 for _ in range(n)] for _ in range(m)]
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1), (0, 0)]
        for i in range(m):
            for j in range(n):
                cnt = 0
                for d in dirs:
                    x = i + d[0]
                    y = j + d[1]
                    if 0 <= x < m and 0 <= y < n:
                        res[i][j] += M[x][y]
                        cnt += 1
                res[i][j] //= cnt
        return res
